Update HTTP status codes

kruai opened the pull request:
# Overview

This PR is being created to address [ESSNTL-1577](https://issues.redhat.com/browse/ESSNTL-1577).
Updates the status codes for our POST/PUT/PATCH/DELETE endpoints to 202, implying that the request has been accepted, rather than completed.

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

ryandillinfelton commented:
I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.

kruai commented:
> I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.

I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?

thearifismail reviewed (commented):
commented

kruai commented on api/host.py:
```diff
@@ -218,7 +218,7 @@ def _delete_filtered_hosts(host_id_list):
 def delete_by_id(host_id_list):
     _delete_filtered_hosts(host_id_list)
-    return flask.Response(None, status.HTTP_200_OK)
+    return flask.Response(None, status.HTTP_202_ACCEPTED)
```
I'm not sure I understand what you're proposing here @thearifismail. Would those changes affect the response status codes?

thearifismail replied:
Response_code should remain 200.  In the above code, `return flask.Response()` is called immediately after `_delete_filtered_host)()`.  What I am suggesting is that `_delete_filtered_hosts()` should be followed by a loop to execute `get_host_by_id()`.  Exit the loop as soon as deletion(s) get confirmed.  If a deleted host is still found, wait for sometime, try again and keep trying for x number of tries .  If that happens then either rollback the change and report the problem or tell the caller to wait for sometime.

kruai replied:
That could work, but we'd have to query both Xjoin and Cyndi to verify that. And we'd have to do that for each endpoint... it honestly gets more complicated the more I think about it :/

ryandillinfelton commented:
> > I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.
> 
> I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?

I don't know that there is a great solution. The best I can think of is to just document the behavior. I don't think it's unreasonable for there to be a short delay before new content is available for querying, and technically a `200` is accurate since the record is deleted instantly in the database. 

Really in hindsight, I would argue we should have versioned the API when we moved to xjoin and kept the DB version available as V1. But since we didn't we effectively introduced a breaking change by changing the level of data consistency that users could expect, directly effecting valid uses of the API despite not *technically* changing the interface.

Whether we return 202 (which would surely break existing business logic) or document the delay, the delay is still there and users still need to wait. If users really need a way to verify the document is modified how they expected we could make a lightweight DB endpoint available that allows only fetching hosts by ID, or tell them to make the request with use `x-data-source: db` header. If they just need the server populated fields like `inventory_id` we should return that in the POST response.

kruai commented:
> > > I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.
> > 
> > 
> > I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?
> 
> I don't know that there is a great solution. The best I can think of is to just document the behavior. I don't think it's unreasonable for there to be a short delay before new content is available for querying, and technically a `200` is accurate since the record is deleted instantly in the database.
> 
> Really in hindsight, I would argue we should have versioned the API when we moved to xjoin and kept the DB version available as V1. But since we didn't we effectively introduced a breaking change by changing the level of data consistency that users could expect, directly effecting valid uses of the API despite not _technically_ changing the interface.
> 
> Whether we return 202 (which would surely break existing business logic) or document the delay, the delay is still there and users still need to wait. If users really need a way to verify the document is modified how they expected we could make a lightweight DB endpoint available that allows only fetching hosts by ID, or tell them to make the request with use `x-data-source: db` header. If they just need the server populated fields like `inventory_id` we should return that in the POST response.



> > > I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.
> > 
> > 
> > I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?
> 
> I don't know that there is a great solution. The best I can think of is to just document the behavior. I don't think it's unreasonable for there to be a short delay before new content is available for querying, and technically a `200` is accurate since the record is deleted instantly in the database.
> 
> Really in hindsight, I would argue we should have versioned the API when we moved to xjoin and kept the DB version available as V1. But since we didn't we effectively introduced a breaking change by changing the level of data consistency that users could expect, directly effecting valid uses of the API despite not _technically_ changing the interface.
> 
> Whether we return 202 (which would surely break existing business logic) or document the delay, the delay is still there and users still need to wait. If users really need a way to verify the document is modified how they expected we could make a lightweight DB endpoint available that allows only fetching hosts by ID, or tell them to make the request with use `x-data-source: db` header. If they just need the server populated fields like `inventory_id` we should return that in the POST response.

Fair point. I think this was brought up because of a situation where someone did a DELETE, and then complained when the host was still in a GET they sent immediately after. Since changing the response statuses will break something, and it would take a lot of work would go into checking Cyndi&Xjoin after every one of these modifications (not to mention the slow responses that would entail), it could be that just documenting the delay would be our best option. This was only going to be a shorter-term change anyway... what do you think, @fstavela ? Tagging you because IIRC you brought the issue to our attention originally

ryandillinfelton commented:
> > > > I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.
> > > 
> > > 
> > > I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?
> > 
> > 
> > I don't know that there is a great solution. The best I can think of is to just document the behavior. I don't think it's unreasonable for there to be a short delay before new content is available for querying, and technically a `200` is accurate since the record is deleted instantly in the database.
> > Really in hindsight, I would argue we should have versioned the API when we moved to xjoin and kept the DB version available as V1. But since we didn't we effectively introduced a breaking change by changing the level of data consistency that users could expect, directly effecting valid uses of the API despite not _technically_ changing the interface.
> > Whether we return 202 (which would surely break existing business logic) or document the delay, the delay is still there and users still need to wait. If users really need a way to verify the document is modified how they expected we could make a lightweight DB endpoint available that allows only fetching hosts by ID, or tell them to make the request with use `x-data-source: db` header. If they just need the server populated fields like `inventory_id` we should return that in the POST response.
> 
> > > > I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.
> > > 
> > > 
> > > I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?
> > 
> > 
> > I don't know that there is a great solution. The best I can think of is to just document the behavior. I don't think it's unreasonable for there to be a short delay before new content is available for querying, and technically a `200` is accurate since the record is deleted instantly in the database.
> > Really in hindsight, I would argue we should have versioned the API when we moved to xjoin and kept the DB version available as V1. But since we didn't we effectively introduced a breaking change by changing the level of data consistency that users could expect, directly effecting valid uses of the API despite not _technically_ changing the interface.
> > Whether we return 202 (which would surely break existing business logic) or document the delay, the delay is still there and users still need to wait. If users really need a way to verify the document is modified how they expected we could make a lightweight DB endpoint available that allows only fetching hosts by ID, or tell them to make the request with use `x-data-source: db` header. If they just need the server populated fields like `inventory_id` we should return that in the POST response.
> 
> Fair point. I think this was brought up because of a situation where someone did a DELETE, and then complained when the host was still in a GET they sent immediately after. Since changing the response statuses will break something, and it would take a lot of work would go into checking Cyndi&Xjoin after every one of these modifications (not to mention the slow responses that would entail), it could be that just documenting the delay would be our best option. This was only going to be a shorter-term change anyway... what do you think, @fstavela ? Tagging you because IIRC you brought the issue to our attention originally

We could also consider including a message in the body of the request telling the user about the delay since no one reads docs.

fstavela commented:
> > > > > I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.
> > > > 
> > > > 
> > > > I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?
> > > 
> > > 
> > > I don't know that there is a great solution. The best I can think of is to just document the behavior. I don't think it's unreasonable for there to be a short delay before new content is available for querying, and technically a `200` is accurate since the record is deleted instantly in the database.
> > > Really in hindsight, I would argue we should have versioned the API when we moved to xjoin and kept the DB version available as V1. But since we didn't we effectively introduced a breaking change by changing the level of data consistency that users could expect, directly effecting valid uses of the API despite not _technically_ changing the interface.
> > > Whether we return 202 (which would surely break existing business logic) or document the delay, the delay is still there and users still need to wait. If users really need a way to verify the document is modified how they expected we could make a lightweight DB endpoint available that allows only fetching hosts by ID, or tell them to make the request with use `x-data-source: db` header. If they just need the server populated fields like `inventory_id` we should return that in the POST response.
> > 
> > 
> > > > > I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.
> > > > 
> > > > 
> > > > I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?
> > > 
> > > 
> > > I don't know that there is a great solution. The best I can think of is to just document the behavior. I don't think it's unreasonable for there to be a short delay before new content is available for querying, and technically a `200` is accurate since the record is deleted instantly in the database.
> > > Really in hindsight, I would argue we should have versioned the API when we moved to xjoin and kept the DB version available as V1. But since we didn't we effectively introduced a breaking change by changing the level of data consistency that users could expect, directly effecting valid uses of the API despite not _technically_ changing the interface.
> > > Whether we return 202 (which would surely break existing business logic) or document the delay, the delay is still there and users still need to wait. If users really need a way to verify the document is modified how they expected we could make a lightweight DB endpoint available that allows only fetching hosts by ID, or tell them to make the request with use `x-data-source: db` header. If they just need the server populated fields like `inventory_id` we should return that in the POST response.
> > 
> > 
> > Fair point. I think this was brought up because of a situation where someone did a DELETE, and then complained when the host was still in a GET they sent immediately after. Since changing the response statuses will break something, and it would take a lot of work would go into checking Cyndi&Xjoin after every one of these modifications (not to mention the slow responses that would entail), it could be that just documenting the delay would be our best option. This was only going to be a shorter-term change anyway... what do you think, @fstavela ? Tagging you because IIRC you brought the issue to our attention originally
> 
> We could also consider including a message in the body of the response telling the user about the delay since no one reads docs.



> > > > > I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.
> > > > 
> > > > 
> > > > I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?
> > > 
> > > 
> > > I don't know that there is a great solution. The best I can think of is to just document the behavior. I don't think it's unreasonable for there to be a short delay before new content is available for querying, and technically a `200` is accurate since the record is deleted instantly in the database.
> > > Really in hindsight, I would argue we should have versioned the API when we moved to xjoin and kept the DB version available as V1. But since we didn't we effectively introduced a breaking change by changing the level of data consistency that users could expect, directly effecting valid uses of the API despite not _technically_ changing the interface.
> > > Whether we return 202 (which would surely break existing business logic) or document the delay, the delay is still there and users still need to wait. If users really need a way to verify the document is modified how they expected we could make a lightweight DB endpoint available that allows only fetching hosts by ID, or tell them to make the request with use `x-data-source: db` header. If they just need the server populated fields like `inventory_id` we should return that in the POST response.
> > 
> > 
> > > > > I don't believe this should be merged. We can't change the response code of existing resources without versioning the API, it's a breaking change.
> > > > 
> > > > 
> > > > I thought we had agreed that updating these responses was the best way to clear up confusion in the short-term... At least, until we implement some way to check that the change has propagated to xjoin and cyndi. What would you suggest?
> > > 
> > > 
> > > I don't know that there is a great solution. The best I can think of is to just document the behavior. I don't think it's unreasonable for there to be a short delay before new content is available for querying, and technically a `200` is accurate since the record is deleted instantly in the database.
> > > Really in hindsight, I would argue we should have versioned the API when we moved to xjoin and kept the DB version available as V1. But since we didn't we effectively introduced a breaking change by changing the level of data consistency that users could expect, directly effecting valid uses of the API despite not _technically_ changing the interface.
> > > Whether we return 202 (which would surely break existing business logic) or document the delay, the delay is still there and users still need to wait. If users really need a way to verify the document is modified how they expected we could make a lightweight DB endpoint available that allows only fetching hosts by ID, or tell them to make the request with use `x-data-source: db` header. If they just need the server populated fields like `inventory_id` we should return that in the POST response.
> > 
> > 
> > Fair point. I think this was brought up because of a situation where someone did a DELETE, and then complained when the host was still in a GET they sent immediately after. Since changing the response statuses will break something, and it would take a lot of work would go into checking Cyndi&Xjoin after every one of these modifications (not to mention the slow responses that would entail), it could be that just documenting the delay would be our best option. This was only going to be a shorter-term change anyway... what do you think, @fstavela ? Tagging you because IIRC you brought the issue to our attention originally
> 
> We could also consider including a message in the body of the response telling the user about the delay since no one reads docs.

I think that waiting and checking that the change has been propagated to xjoin and cyndi and returning 200 after it's done is the ultimate goal here. However, we should consider that this will not affect a lot of customers (majority probably wouldn't even notice), so I'd say that it is probably a very low priority change and if it requires a lot of work, it's probably not worth it.

If changing the response code would break anything (I didn't think of that when we were discussing this change), I think that updating the docs and including a message in the response body is enough as a temporary solution.

kruai commented:
OK. I created [ESSNTL-2005](https://issues.redhat.com/browse/ESSNTL-2005) to update our docs, and I may update the response body for these endpoints as well.
